File: app/src/main/java/de/christinecoenen/code/zapp/models/channels/json/SortableJsonChannelList.java
package de.christinecoenen.code.zapp.models.channels.json;


import android.content.Context;
import androidx.annotation.NonNull;

import java.util.Iterator;
import java.util.List;

import de.christinecoenen.code.zapp.models.channels.ChannelModel;
import de.christinecoenen.code.zapp.models.channels.IChannelList;
import de.christinecoenen.code.zapp.models.channels.ISortableChannelList;
import de.christinecoenen.code.zapp.models.channels.SimpleChannelList;
import de.christinecoenen.code.zapp.app.settings.helper.PreferenceChannelOrderHelper;

public class SortableJsonChannelList implements ISortableChannelList {

	final Context context;
	final PreferenceChannelOrderHelper channelOrderHelper;
	IChannelList channelList;

	public SortableJsonChannelList(Context context) {
		this.context = context;
		channelOrderHelper = new PreferenceChannelOrderHelper(context);
		reload();
	}

	@Override
	public void reload() {
		loadSortingFromDisk();
	}

	@Override
	public ChannelModel get(int index) {
		return channelList.get(index);
	}

	@Override
	public ChannelModel get(String id) {
		return channelList.get(id);
	}

	@Override
	public int size() {
		return channelList.size();
	}

	@Override
	public List<ChannelModel> getList() {
		return channelList.getList();
	}

	@NonNull
	@Override
	public Iterator<ChannelModel> iterator() {
		return channelList.iterator();
	}

	@Override
	public int indexOf(String channelId) {
		return channelList.indexOf(channelId);
	}

	@Override
	public void reloadChannelOrder() {
		loadSortingFromDisk();
	}

	@Override
	public void persistChannelOrder() {
		channelOrderHelper.saveChannelOrder(channelList.getList());
	}

	void loadSortingFromDisk() {
		List<ChannelModel> listFromDisk = new JsonChannelList(context).getList();
		List<ChannelModel> sortedChannels = channelOrderHelper.sortChannelList(listFromDisk, false);
		channelList = new SimpleChannelList(sortedChannels);
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/models/channels/json/SortableJsonChannelList.kt
package de.christinecoenen.code.zapp.models.channels.json

import android.content.Context
import de.christinecoenen.code.zapp.app.settings.helper.PreferenceChannelOrderHelper
import de.christinecoenen.code.zapp.models.channels.ChannelModel
import de.christinecoenen.code.zapp.models.channels.IChannelList
import de.christinecoenen.code.zapp.models.channels.ISortableChannelList
import de.christinecoenen.code.zapp.models.channels.SimpleChannelList

open class SortableJsonChannelList(val context: Context) : ISortableChannelList {

	val channelOrderHelper: PreferenceChannelOrderHelper = PreferenceChannelOrderHelper(context)
	var channelList: IChannelList = SimpleChannelList(listOf())

	init {
		reload()
	}

	override fun get(index: Int) = channelList[index]

	override fun get(id: String) = channelList[id]

	override fun size() = channelList.size()

	override val list = channelList.list

	override fun iterator() = channelList.iterator()

	override fun indexOf(channelId: String) = channelList.indexOf(channelId)

	final override fun reload() {
		loadSortingFromDisk()
	}

	override fun reloadChannelOrder() {
		loadSortingFromDisk()
	}

	override fun persistChannelOrder() {
		channelOrderHelper.saveChannelOrder(channelList.list)
	}

	open fun loadSortingFromDisk() {
		val listFromDisk: List<ChannelModel> = JsonChannelList(context).list
		val sortedChannels = channelOrderHelper.sortChannelList(listFromDisk, false)

		channelList = SimpleChannelList(sortedChannels)
	}
}
